### ISSUE TITLE
Composer breaks installation on require

### ISSUE THREAD

**Issue by IchHabRecht** [composer/composer#6838]
When I run these commands:

```
composer require typo3/cms="^7.6"
composer config repositories.typo3 composer https://composer.typo3.org
composer require typo3-ter/jumpurl
composer require friendsoftypo3/jumpurl
```

Everything is required and installed properly. But running `composer install` afterwards or on a completely different system outputs an error:

```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Conclusion: remove friendsoftypo3/jumpurl 7.7.0
    - don't install friendsoftypo3/jumpurl 7.7.0|remove typo3-ter/jumpurl 7.7.0
    - don't install typo3-ter/jumpurl 7.7.0|don't install friendsoftypo3/jumpurl 7.7.0
    - Installation request for friendsoftypo3/jumpurl 7.7.0 -> satisfiable by friendsoftypo3/jumpurl[7.7.0].
    - Installation request for typo3-ter/jumpurl 7.7.0 -> satisfiable by typo3-ter/jumpurl[7.7.0].
```

I think it is because the master version of friendsoftypo3/jumpurl contains a replacement for typo3-ter/jumpurl. But the current locked versions do not have any replacement for themselves. Is this a valid error? And if so, why can I require both packages? Shouldn't the error occur on require instead of install then?

**Comment by alcohol** [comment]
```
rob@macbookpro.local /tmp/test > composer require typo3-ter/jumpurl
Config option "cms-package-dir" has not been set or set to a value different from "{$vendor-dir}/typo3/cms".
 This option will be removed without substitution with typo3/cms-composer-installers 2.0.
 With 2.0 the typo3/cms package will always be installed in the vendor directory.
 To get rid of this warning, use the following command to set the option to a not deprecated value:
 composer config extra.typo3/cms.cms-package-dir '{$vendor-dir}/typo3/cms'
Using version ^7.7 for typo3-ter/jumpurl
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing typo3-ter/jumpurl (7.7.0): Downloading (100%)
Writing lock file
Generating autoload files
Setting web-dir to the composer root directory is highly discouraged for security reasons.
 The default value for this option will change to "web" as of typo3/cms-composer-installers 2.0.
```
```
rob@macbookpro.local /tmp/test > composer require friendsoftypo3/jumpurl
Config option "cms-package-dir" has not been set or set to a value different from "{$vendor-dir}/typo3/cms".
 This option will be removed without substitution with typo3/cms-composer-installers 2.0.
 With 2.0 the typo3/cms package will always be installed in the vendor directory.
 To get rid of this warning, use the following command to set the option to a not deprecated value:
 composer config extra.typo3/cms.cms-package-dir '{$vendor-dir}/typo3/cms'
Using version ^7.7 for friendsoftypo3/jumpurl
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 1 install, 1 update, 0 removals
  - Updating typo3-ter/jumpurl (7.7.0 => 7.7.0):   - Installing typo3-ter/jumpurl (7.7.0): Loading from cache

  - Installing friendsoftypo3/jumpurl (7.7.0): Downloading (100%)
Writing lock file
Generating autoload files
Setting web-dir to the composer root directory is highly discouraged for security reasons.
 The default value for this option will change to "web" as of typo3/cms-composer-installers 2.0.
```

From a user perspective, it would make more sense to first run `composer remove typo3-ter/jumpurl` and then run `composer require friendsoftypo3/jumpurl` since you know the latter replaces the former. So it seems you have discovered an edge case for a scenario that does not really make sense and thus thankfully almost never occurs.

@naderman peculiar though that in the second `require` action it re-installs the first package again. Any idea what could be going wrong here?

**Comment by IchHabRecht** [comment]
Hi @alcohol,

> From a user perspective, it would make more sense to first run composer remove typo3-ter/jumpurl and then run composer require friendsoftypo3/jumpurl since you know the latter replaces the former. So it seems you have discovered an edge case for a scenario that does not really make sense and thus thankfully almost never occurs.

I didn't know the later replaces the former. Just found out while trying to solve the issue, that the master has the replace part. The installed version doesn't replace anything.

**Comment by alcohol** [comment]
```
rob@macbookpro.local /tmp/test > composer show friendsoftypo3/jumpurl
Config option "cms-package-dir" has not been set or set to a value different from "{$vendor-dir}/typo3/cms".
 This option will be removed without substitution with typo3/cms-composer-installers 2.0.
 With 2.0 the typo3/cms package will always be installed in the vendor directory.
 To get rid of this warning, use the following command to set the option to a not deprecated value:
 composer config extra.typo3/cms.cms-package-dir '{$vendor-dir}/typo3/cms'
name     : friendsoftypo3/jumpurl
descrip. : TYPO3 Jump URL handling
keywords :
versions : * 7.7.0
type     : typo3-cms-extension
license  : GPL-2.0+
source   : [git] https://github.com/FriendsOfTYPO3/jumpurl.git 224aa6899ba92f852d3ca819e0199d11c820b4a1
dist     : [zip] https://api.github.com/repos/FriendsOfTYPO3/jumpurl/zipball/224aa6899ba92f852d3ca819e0199d11c820b4a1 224aa6899ba92f852d3ca819e0199d11c820b4a1
names    : friendsoftypo3/jumpurl, jumpurl

autoload
psr-4
FoT3\Jumpurl\ => Classes/

requires
typo3/cms-core ^7.6 || ^8.0
typo3/cms-frontend ^7.6 || ^8.0

replaces
jumpurl *
```

It's right in the details there. Although the fact that they share the same name also kind of gives it away already (or at least should make you curious I imagine).

Anyway, does not matter. We can look into why the replace is not really behaving as one might expect (or why we don't fail more gracefully).

**Comment by IchHabRecht** [comment]
Hi @alcohol,

> It's right in the details there. Although the fact that they share the same name also kind of gives it away already (or at least should make you curious I imagine).

I have no clue, what you mean. I don't expect some package that replaces "jumpurl" to also replace "friendsoftypo3/jumpurl" and/or "typo3-ter/jumpurl". I can try to rebuild the problem with two packages having completely different names. I guess the main issue still stays: that composer takes replacements of the wrong branch and breaks `composer install`

**Comment by alcohol** [comment]
I don't really follow. Do you just randomly include packages without reading their readme or their `composer.json`? I personally like to know what I am using.

Anyway, it does not matter. Pointless discussion. Your issue is noted and we'll look at it.

**Comment by IchHabRecht** [comment]
@alcohol 
I included typo3-ter/jumpurl in my composer.json as dependency, another dependency of mine changed their dependency to friendsoftypo3/jumpurl. Nothing I have done here. Just ran into the error and found it worth to report. Please don't get irrelevant and be thankful for report. Thanks!

**Comment by Seldaek** [comment]
Fixed by 2.0 IMO, the last step from your original post fails here with:

```
  Problem 1
    - typo3-ter/jumpurl is locked to version 7.8.0 and an update of this package was not requested.
    - friendsoftypo3/jumpurl 7.8.0 can not be installed as that would require removing typo3-ter/jumpurl 7.8.0. They both replace jumpurl and can thus not coexist.
    - Root composer.json requires friendsoftypo3/jumpurl ^7.8 -> satisfiable by friendsoftypo3/jumpurl[7.8.0].
```

Which explains hopefully clearly that you need to whitelist update typo3-ter/jumpurl too so it can be removed. i.e.:

```
composer require friendsoftypo3/jumpurl --no-update
composer update friendsoftypo3/jumpurl typo3-ter/jumpurl
```

Then things are in a good state and an install from a clean vendor dir seems to work.
